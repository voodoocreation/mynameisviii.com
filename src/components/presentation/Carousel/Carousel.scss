@import "../../../scss/variables", "../../../scss/mixins";

.Carousel {
  overflow: hidden;

  &--slides {
    line-height: 0;
    white-space: nowrap;
    transform: translate3d(0, 0, 0);
    transition: transform 200ms ease-out;

    > * {
      display: inline-block;
      line-height: normal;
    }
  }

  &--pagination {
    line-height: 0;
    text-align: center;
    padding: $spacing-half 0;

    &--page {
      $size: 24px;

      width: $size;
      height: $size;
      margin: 0 ($spacing / 5);
      border: $text-color 1px solid;
      border-radius: 100%;

      span {
        @include hide-visually();
      }

      &.isSelected {
        background: $text-color;
      }
    }
  }
}
