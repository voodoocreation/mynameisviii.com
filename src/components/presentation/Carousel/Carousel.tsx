import cn from "classnames";
import * as React from "react";

import Button from "../Button/Button";
import "./Carousel.scss";

interface IProps {
  className?: string;
  currentIndex?: number;
  onSlideChange: (index: number) => void;
}

interface IState {
  currentIndex: number;
}

class Carousel extends React.Component<IProps, IState> {
  public static defaultProps = {
    currentIndex: 0
  };

  constructor(props: IProps) {
    super(props);

    this.state = {
      currentIndex: props.currentIndex || 0
    };
  }

  public componentDidUpdate(_: IProps, prevState: IState) {
    if (
      this.props.currentIndex !== undefined &&
      this.props.currentIndex !== prevState.currentIndex
    ) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        currentIndex: this.props.currentIndex
      });
    }
  }

  public render() {
    const { children, className } = this.props;

    return (
      <div className={cn("Carousel", className)}>
        <div
          className="Carousel--slides"
          style={{
            transform: `translate3d(-${this.state.currentIndex * 100}%, 0, 0)`
          }}
        >
          {children}
        </div>
        {this.renderPagination()}
      </div>
    );
  }

  private onPaginationClick = (index: number) => () => {
    this.setState({
      currentIndex: index
    });

    this.props.onSlideChange(index);
  };

  private renderPagination() {
    const { children } = this.props;

    if (!children || React.Children.count(children) < 2) {
      return null;
    }

    return (
      <div className="Carousel--pagination">
        {React.Children.map(children, (_, index) => (
          <Button
            className={cn("Carousel--pagination--page", {
              isSelected: index === this.state.currentIndex
            })}
            isStyled={false}
            onClick={this.onPaginationClick(index)}
          >
            <span>{index + 1}</span>
          </Button>
        ))}
      </div>
    );
  }
}

export default Carousel;
