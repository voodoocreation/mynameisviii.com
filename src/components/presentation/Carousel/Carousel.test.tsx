import * as React from "react";

import WrapperWithIntl from "../../../utilities/WrapperWithIntl";
import Carousel from "./Carousel";

const component = new WrapperWithIntl(Carousel).withDefaultProps({
  onSlideChange: jest.fn()
});

describe("[presentation] <Carousel />", () => {
  describe("when there are several children", () => {
    const wrapper = component
      .withChildren([
        <div className="Slide" key="slide-1" />,
        <div className="Slide" key="slide-2" />,
        <div className="Slide" key="slide-3" />
      ])
      .mount();

    it("renders all of the children", () => {
      expect(wrapper.find(".Slide")).toHaveLength(3);
    });

    it("renders the pagination correctly", () => {
      expect(wrapper.find(".Carousel--pagination")).toHaveLength(1);
      expect(wrapper.find(".Carousel--pagination Button")).toHaveLength(3);
    });

    it("clicks a pagination page", () => {
      wrapper
        .find(".Carousel--pagination Button")
        .at(1)
        .simulate("click");
    });

    it("calls onSlideChange prop with expected payload", () => {
      expect(component.props.onSlideChange).toHaveBeenCalledTimes(1);
      expect(component.props.onSlideChange).toHaveBeenCalledWith(1);
    });

    it("changes the current slide", () => {
      // @ts-ignore-next-line
      expect(wrapper.find(".Carousel--slides").prop("style").transform).toBe(
        "translate3d(-100%, 0, 0)"
      );
      expect(
        wrapper
          .find(".Carousel--pagination Button")
          .at(1)
          .hasClass("isSelected")
      ).toBe(true);
    });

    it("receives new props", () => {
      wrapper.setProps({ currentIndex: 2 }).update();
    });

    it("changes the current slide", () => {
      // @ts-ignore-next-line
      expect(wrapper.find(".Carousel--slides").prop("style").transform).toBe(
        "translate3d(-200%, 0, 0)"
      );
      expect(
        wrapper
          .find(".Carousel--pagination Button")
          .at(2)
          .hasClass("isSelected")
      ).toBe(true);
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  describe("when there's only one child", () => {
    const wrapper = component
      .withChildren([<div className="Slide" key="slide-1" />])
      .render();

    it("renders all of the children", () => {
      expect(wrapper.find(".Slide")).toHaveLength(1);
    });

    it("doesn't render the pagination", () => {
      expect(wrapper.find(".Carousel--pagination")).toHaveLength(0);
    });

    it("matches snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
