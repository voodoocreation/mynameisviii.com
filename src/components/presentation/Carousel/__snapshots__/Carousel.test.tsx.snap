// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <Carousel /> when there are several children matches snapshot 1`] = `
<div
  class="Carousel"
>
  <div
    class="Carousel--slides"
    style="transform: translate3d(-200%, 0, 0);"
  >
    <div
      class="Slide"
    />
    <div
      class="Slide"
    />
    <div
      class="Slide"
    />
  </div>
  <div
    class="Carousel--pagination"
  >
    <button
      class="Button Carousel--pagination--page"
      type="button"
    >
      <span
        class="Button--content"
      >
        <span>
          1
        </span>
      </span>
    </button>
    <button
      class="Button Carousel--pagination--page"
      type="button"
    >
      <span
        class="Button--content"
      >
        <span>
          2
        </span>
      </span>
    </button>
    <button
      class="Button Carousel--pagination--page isSelected"
      type="button"
    >
      <span
        class="Button--content"
      >
        <span>
          3
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`[presentation] <Carousel /> when there's only one child matches snapshot 1`] = `
<div
  class="Carousel"
>
  <div
    class="Carousel--slides"
    style="transform:translate3d(-0%, 0, 0)"
  >
    <div
      class="Slide"
    />
  </div>
</div>
`;
